File: android/app/src/main/java/com/anonymous/todolistapp/TodoWidgetProvider.java
// android/app/src/main/java/com/anonymous/todolistapp/TodoWidgetProvider.java
// 🔧 최소한의 테스트 위젯
package com.anonymous.todolistapp;

import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.Context;
import android.widget.RemoteViews;
import android.util.Log;
import org.json.JSONArray;
import org.json.JSONObject;

public class TodoWidgetProvider extends AppWidgetProvider {
    
    private static final String TAG = "TodoWidget";

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        Log.d(TAG, "onUpdate called with " + appWidgetIds.length + " widgets");
        
        for (int appWidgetId : appWidgetIds) {
            updateAppWidget(context, appWidgetManager, appWidgetId);
        }
    }

    static void updateAppWidget(Context context, AppWidgetManager appWidgetManager, int appWidgetId) {
        Log.d(TAG, "updateAppWidget called for widget ID: " + appWidgetId);
        
        try {
            RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.simple_widget);
            
            // 🎮 실제 할 일 데이터 로드
            JSONArray todos = loadTodos(context);
            JSONObject firstTodo = findFirstActiveTodo(todos);
            
            if (firstTodo != null) {
                String title = firstTodo.getString("title");
                int timeLeft = firstTodo.getInt("timeLeft");
                boolean completed = firstTodo.getBoolean("completed");
                
                // 위젯 제목 표시
                views.setViewVisibility(R.id.widget_title, android.view.View.VISIBLE);
                views.setTextViewText(R.id.widget_title, "할 일 목록");
                
                // 할 일 아이템 표시
                views.setViewVisibility(R.id.todo_item_container, android.view.View.VISIBLE);
                views.setViewVisibility(R.id.simple_text, android.view.View.GONE);
                
                // 제목 표시 (10글자 제한 - React Native와 동일)
                String truncatedTitle = truncateTitle(title, 10);
                views.setTextViewText(R.id.todo_title, truncatedTitle);
                
                // 🎮 React Native와 동일한 21개 체력바 로직!
                renderHealthBars(views, timeLeft, completed);
                
                Log.d(TAG, "Showing todo: " + title + " (timeLeft: " + timeLeft + ", completed: " + completed + ")");
            } else {
                // 빈 상태 표시
                views.setViewVisibility(R.id.widget_title, android.view.View.VISIBLE);
                views.setTextViewText(R.id.widget_title, "할 일 목록");
                views.setViewVisibility(R.id.todo_item_container, android.view.View.GONE);
                views.setViewVisibility(R.id.simple_text, android.view.View.VISIBLE);
                views.setTextViewText(R.id.simple_text, "할 일이 없습니다\n앱에서 할 일을 추가하세요!");
            }
            
            // 위젯 업데이트
            appWidgetManager.updateAppWidget(appWidgetId, views);
            
            Log.d(TAG, "Widget updated successfully");
            
        } catch (Exception e) {
            Log.e(TAG, "Error updating widget: " + e.getMessage(), e);
            
            // 에러 시 기본 메시지
            RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.simple_widget);
            views.setViewVisibility(R.id.simple_text, android.view.View.VISIBLE);
            views.setTextViewText(R.id.simple_text, "위젯 에러\n" + e.getMessage());
            appWidgetManager.updateAppWidget(appWidgetId, views);
        }
    }
    
    @Override
    public void onEnabled(Context context) {
        Log.d(TAG, "Widget enabled");
    }

    @Override
    public void onDisabled(Context context) {
        Log.d(TAG, "Widget disabled");
    }
    
    // 🔧 WidgetDataManager에서 호출하는 메서드 추가
    public static void updateWidgetData(Context context, JSONArray todos) {
        Log.d(TAG, "updateWidgetData called with " + todos.length() + " todos");
        
        try {
            // 위젯 업데이트 (현재는 단순히 새로고침)
            AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(context);
            android.content.ComponentName thisWidget = new android.content.ComponentName(context, TodoWidgetProvider.class);
            int[] appWidgetIds = appWidgetManager.getAppWidgetIds(thisWidget);
            
            for (int appWidgetId : appWidgetIds) {
                updateAppWidget(context, appWidgetManager, appWidgetId);
            }
            
            Log.d(TAG, "Widget data updated successfully");
        } catch (Exception e) {
            Log.e(TAG, "Error updating widget data: " + e.getMessage(), e);
        }
    }
    
    // 🎮 할 일 데이터 로드
    private static JSONArray loadTodos(Context context) {
        try {
            android.content.SharedPreferences prefs = context.getSharedPreferences("TodoWidgetPrefs", Context.MODE_PRIVATE);
            String todosJson = prefs.getString("todos", "[]");
            Log.d(TAG, "Loaded todos JSON: " + todosJson);
            return new JSONArray(todosJson);
        } catch (Exception e) {
            Log.e(TAG, "Error loading todos: " + e.getMessage(), e);
            return new JSONArray();
        }
    }
    
    // 첫 번째 활성 할 일 찾기
    private static JSONObject findFirstActiveTodo(JSONArray todos) {
        for (int i = 0; i < todos.length(); i++) {
            try {
                JSONObject todo = todos.getJSONObject(i);
                if (!todo.getBoolean("completed")) {
                    return todo;
                }
            } catch (Exception e) {
                Log.e(TAG, "Error checking todo: " + e.getMessage(), e);
            }
        }
        return null;
    }
    
    // 제목 자르기 (React Native와 동일 - 10글자)
    private static String truncateTitle(String title, int maxLength) {
        if (title == null) return "";
        if (title.length() <= maxLength) {
            return title;
        }
        return title.substring(0, maxLength) + "...";
    }
    
    // 🎮 React Native와 동일한 21개 체력바 렌더링!
    private static void renderHealthBars(RemoteViews views, int timeLeft, boolean completed) {
        final int maxBars = 21; // 3주 (React Native와 동일)
        
        // 바의 개수 계산 (React Native 로직과 완전 동일)
        int barCount;
        if (completed) {
            barCount = maxBars; // 완료된 경우 모든 바 표시 (회색으로)
        } else if (timeLeft < 0) {
            barCount = 0; // 지난 날짜인 경우 0개
        } else {
            barCount = Math.min(Math.max(timeLeft, 0), maxBars);
        }
        
        // 색상 결정 (React Native getBarColor 함수와 동일)
        int activeColor = getBarColor(timeLeft, completed);
        int inactiveColor = 0xFFE0E0E0; // 비활성 바 색상
        
        // 21개 바의 ID 배열
        int[] barIds = {
            R.id.bar_1, R.id.bar_2, R.id.bar_3, R.id.bar_4, R.id.bar_5, R.id.bar_6, R.id.bar_7,
            R.id.bar_8, R.id.bar_9, R.id.bar_10, R.id.bar_11, R.id.bar_12, R.id.bar_13, R.id.bar_14,
            R.id.bar_15, R.id.bar_16, R.id.bar_17, R.id.bar_18, R.id.bar_19, R.id.bar_20, R.id.bar_21
        };
        
        // 각 바의 색상 설정 (React Native 로직과 동일)
        for (int i = 0; i < barIds.length; i++) {
            boolean isActive;
            if (completed) {
                isActive = true; // 완료된 경우 모든 바를 활성으로 (색상은 회색)
            } else {
                isActive = i < barCount;
            }
            
            int color = isActive ? activeColor : inactiveColor;
            views.setInt(barIds[i], "setBackgroundColor", color);
        }
        
        Log.d(TAG, "Health bars rendered: barCount=" + barCount + ", completed=" + completed + ", timeLeft=" + timeLeft);
    }
    
    // 🎮 React Native getBarColor 함수와 완전 동일!
    private static int getBarColor(int timeLeft, boolean completed) {
        if (completed) {
            return 0xFFD0D0D0; // 회색 - 완료됨 (React Native와 동일)
        } else if (timeLeft < 0) {
            return 0xFFD32F2F; // 빨간색 - 지난 일
        } else if (timeLeft == 0) {
            return 0xFFFF9800; // 주황색 - 오늘
        } else if (timeLeft <= 3) {
            return 0xFFF57C00; // 주황색 - 3일 이내
        } else {
            return 0xFF388E3C; // 초록색 - 여유있음
        }
    }
}

File: android/app/src/main/java/com/anonymous/todolistapp/TodoWidgetProvider.kt
// android/app/src/main/java/com/anonymous/todolistapp/TodoWidgetProvider.kt
package com.anonymous.todolistapp

import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.Context
import android.content.Intent
import android.graphics.*
import android.util.Log
import android.widget.RemoteViews
import org.json.JSONArray
import org.json.JSONObject
import java.util.*

class TodoWidgetProvider : AppWidgetProvider() {
    
    companion object {
        private const val TAG = "TodoWidget"
        private const val MAX_BARS = 21 // 3주
        
        // 버튼 액션들
        private const val ACTION_COMPLETE_TODO = "com.anonymous.todolistapp.ACTION_COMPLETE_TODO"
        private const val ACTION_DELETE_TODO = "com.anonymous.todolistapp.ACTION_DELETE_TODO"
        private const val ACTION_OPEN_APP = "com.anonymous.todolistapp.ACTION_OPEN_APP"
        private const val EXTRA_TODO_ID = "todo_id"
        private const val EXTRA_CATEGORY_ID = "category_id"
        
        fun updateWidgetData(context: Context, todos: JSONArray) {
            Log.d(TAG, "updateWidgetData called with ${todos.length()} todos")
            
            try {
                val appWidgetManager = AppWidgetManager.getInstance(context)
                val thisWidget = android.content.ComponentName(context, TodoWidgetProvider::class.java)
                val appWidgetIds = appWidgetManager.getAppWidgetIds(thisWidget)
                
                // 리스트뷰 데이터 새로고침
                appWidgetManager.notifyAppWidgetViewDataChanged(appWidgetIds, R.id.todo_list)
                
                for (appWidgetId in appWidgetIds) {
                    updateAppWidget(context, appWidgetManager, appWidgetId)
                }
                
                Log.d(TAG, "Widget data updated successfully")
            } catch (e: Exception) {
                Log.e(TAG, "Error updating widget data: ${e.message}", e)
            }
        }
        
        private fun updateAppWidget(context: Context, appWidgetManager: AppWidgetManager, appWidgetId: Int) {
            Log.d(TAG, "updateAppWidget called for widget ID: $appWidgetId")
            
            try {
                val views = RemoteViews(context.packageName, R.layout.todo_widget_scrollable_layout)
                
                // 리스트뷰 어댑터 설정
                val serviceIntent = Intent(context, TodoWidgetService::class.java)
                views.setRemoteAdapter(R.id.todo_list, serviceIntent)
                
                // 빈 뷰 설정
                views.setEmptyView(R.id.todo_list, android.R.id.empty)
                
                // 버튼 클릭을 위한 인텐트 템플릿 설정
                val buttonIntent = Intent(context, TodoWidgetProvider::class.java)
                val buttonPendingIntent = android.app.PendingIntent.getBroadcast(
                    context, 0, buttonIntent, 
                    android.app.PendingIntent.FLAG_UPDATE_CURRENT or android.app.PendingIntent.FLAG_MUTABLE
                )
                views.setPendingIntentTemplate(R.id.todo_list, buttonPendingIntent)
                
                appWidgetManager.updateAppWidget(appWidgetId, views)
                Log.d(TAG, "Widget updated successfully")
                
            } catch (e: Exception) {
                Log.e(TAG, "Error updating widget: ${e.message}", e)
                
                // 에러 시 기본 메시지
                val views = RemoteViews(context.packageName, R.layout.todo_widget_scrollable_layout)
                views.setTextViewText(android.R.id.empty, "위젯 에러: ${e.message}")
                appWidgetManager.updateAppWidget(appWidgetId, views)
            }
        }
        
        // 할 일 데이터 로드
        private fun loadTodos(context: Context): JSONArray {
            return try {
                val prefs = context.getSharedPreferences("TodoWidgetPrefs", Context.MODE_PRIVATE)
                val todosJson = prefs.getString("todos", "[]") ?: "[]"
                Log.d(TAG, "Loaded todos JSON: $todosJson")
                JSONArray(todosJson)
            } catch (e: Exception) {
                Log.e(TAG, "Error loading todos: ${e.message}", e)
                JSONArray()
            }
        }
        
        // 다음 루틴 날짜 계산 (완료 시점부터)
        private fun calculateNextRoutineDate(routineType: String, routineConfig: JSONObject): Int {
            val now = Calendar.getInstance()
            now.set(Calendar.HOUR_OF_DAY, 0)
            now.set(Calendar.MINUTE, 0)
            now.set(Calendar.SECOND, 0)
            now.set(Calendar.MILLISECOND, 0)
            
            return when (routineType) {
                "daily" -> 1 // 내일
                "weekly" -> {
                    val days = routineConfig.optJSONArray("days")
                    if (days != null) {
                        val todayDay = now.get(Calendar.DAY_OF_WEEK) - 1 // 0=일요일, 1=월요일...
                        val selectedDays = mutableListOf<Int>()
                        
                        for (i in 0 until days.length()) {
                            selectedDays.add(days.getInt(i))
                        }
                        
                        selectedDays.sort()
                        
                        // 완료 시점(오늘) 이후의 가장 가까운 요일 찾기
                        var nextDay = selectedDays.find { it > todayDay }
                        if (nextDay == null) {
                            nextDay = selectedDays[0] // 다음 주의 첫 번째 요일
                        }
                        
                        val daysUntilNext = if (nextDay > todayDay) {
                            nextDay - todayDay
                        } else {
                            7 - todayDay + nextDay
                        }
                        
                        daysUntilNext
                    } else {
                        7 // 기본값: 일주일 후
                    }
                }
                "monthly" -> {
                    val date = routineConfig.optInt("date", 1)
                    val currentMonth = now.get(Calendar.MONTH)
                    val currentYear = now.get(Calendar.YEAR)
                    val currentDate = now.get(Calendar.DAY_OF_MONTH)
                    
                    var nextMonth = currentMonth
                    var nextYear = currentYear
                    
                    // 완료 시점(오늘) 이후의 해당 날짜로 설정
                    if (currentDate >= date) {
                        nextMonth += 1
                        if (nextMonth > 11) {
                            nextMonth = 0
                            nextYear += 1
                        }
                    }
                    
                    val nextDate = Calendar.getInstance()
                    nextDate.set(nextYear, nextMonth, date, 0, 0, 0)
                    nextDate.set(Calendar.MILLISECOND, 0)
                    
                    val timeDiff = nextDate.timeInMillis - now.timeInMillis
                    Math.ceil(timeDiff / (1000.0 * 60 * 60 * 24)).toInt()
                }
                else -> 1
            }
        }
    }
    
    override fun onReceive(context: Context, intent: Intent) {
        super.onReceive(context, intent)
        
        when (intent.action) {
            ACTION_COMPLETE_TODO -> {
                val todoId = intent.getStringExtra(EXTRA_TODO_ID)
                val categoryId = intent.getStringExtra(EXTRA_CATEGORY_ID)
                if (todoId != null && categoryId != null) {
                    handleCompleteTodo(context, categoryId, todoId)
                }
            }
            ACTION_DELETE_TODO -> {
                val todoId = intent.getStringExtra(EXTRA_TODO_ID)
                val categoryId = intent.getStringExtra(EXTRA_CATEGORY_ID)
                if (todoId != null && categoryId != null) {
                    handleDeleteTodo(context, categoryId, todoId)
                }
            }
            ACTION_OPEN_APP -> {
                handleOpenApp(context)
            }
        }
    }
    
    private fun handleCompleteTodo(context: Context, categoryId: String, todoId: String) {
        try {
            // SharedPreferences에서 데이터 로드
            val prefs = context.getSharedPreferences("TodoWidgetPrefs", Context.MODE_PRIVATE)
            val todosJson = prefs.getString("todos", "[]") ?: "[]"
            val todos = JSONArray(todosJson)
            
            // 해당 할 일 찾아서 처리
            var found = false
            for (i in 0 until todos.length()) {
                val todo = todos.getJSONObject(i)
                if (todo.getString("id") == todoId) {
                    val isRoutine = todo.optBoolean("isRoutine", false)
                    val currentCompleted = todo.optBoolean("completed", false)
                    
                    if (isRoutine) {
                        if (currentCompleted) {
                            // 루틴 업무가 이미 완료된 경우 - 미완료로 되돌리기
                            todo.put("completed", false)
                            Log.d(TAG, "Routine todo uncompleted: $todoId")
                        } else {
                            // 루틴 업무 완료 처리
                            val routineType = todo.optString("routineType", "daily")
                            val routineConfig = todo.optJSONObject("routineConfig") ?: JSONObject()
                            
                            val nextTimeLeft = calculateNextRoutineDate(routineType, routineConfig)
                            todo.put("timeLeft", nextTimeLeft)
                            todo.put("completed", true) // 루틴 완료 시 일시적으로 완료 상태로 표시
                            
                            Log.d(TAG, "Routine todo completed: $todoId, nextTimeLeft: $nextTimeLeft")
                        }
                        
                        // 위젯 변경사항 기록
                        recordWidgetChange(context, "TOGGLE_COMPLETED", categoryId, todoId)
                    } else {
                        // 일반 업무인 경우: 완료 처리하고 위젯에서 제거
                        todo.put("completed", true)
                        
                        Log.d(TAG, "Regular todo completed: $todoId")
                        
                        // 위젯 변경사항 기록
                        recordWidgetChange(context, "TOGGLE_COMPLETED", categoryId, todoId)
                    }
                    
                    found = true
                    break
                }
            }
            
            if (found) {
                // SharedPreferences에 저장
                prefs.edit().apply {
                    putString("todos", todos.toString())
                    apply()
                }
                
                // 위젯 업데이트
                updateWidgetData(context, todos)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error completing todo: ${e.message}", e)
        }
    }
    
    private fun handleDeleteTodo(context: Context, categoryId: String, todoId: String) {
        try {
            // SharedPreferences에서 데이터 로드
            val prefs = context.getSharedPreferences("TodoWidgetPrefs", Context.MODE_PRIVATE)
            val todosJson = prefs.getString("todos", "[]") ?: "[]"
            val todos = JSONArray(todosJson)
            
            // 해당 할 일 찾아서 삭제
            var foundIndex = -1
            for (i in 0 until todos.length()) {
                val todo = todos.getJSONObject(i)
                if (todo.getString("id") == todoId) {
                    foundIndex = i
                    break
                }
            }
            
            if (foundIndex != -1) {
                todos.remove(foundIndex)
                
                // SharedPreferences에 저장
                prefs.edit().apply {
                    putString("todos", todos.toString())
                    apply()
                }
                
                // 위젯 변경사항 기록
                recordWidgetChange(context, "DELETE", categoryId, todoId)
                
                // 위젯 업데이트
                updateWidgetData(context, todos)
                
                Log.d(TAG, "Todo deleted: $todoId")
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error deleting todo: ${e.message}", e)
        }
    }
    
    private fun handleOpenApp(context: Context) {
        try {
            val packageManager = context.packageManager
            val intent = packageManager.getLaunchIntentForPackage(context.packageName)
            intent?.let {
                it.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                context.startActivity(it)
                Log.d(TAG, "App opened from widget")
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error opening app: ${e.message}", e)
        }
    }
    
    private fun recordWidgetChange(context: Context, action: String, categoryId: String, todoId: String) {
        try {
            val prefs = context.getSharedPreferences("TodoWidgetPrefs", Context.MODE_PRIVATE)
            val existingChanges = prefs.getString("widgetChanges", "[]") ?: "[]"
            val changes = JSONArray(existingChanges)
            
            val change = JSONObject().apply {
                put("action", action)
                put("categoryId", categoryId)
                put("todoId", todoId)
                put("timestamp", System.currentTimeMillis())
            }
            
            changes.put(change)
            
            prefs.edit().apply {
                putString("widgetChanges", changes.toString())
                apply()
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error recording widget change: ${e.message}", e)
        }
    }
    
    override fun onEnabled(context: Context) {
        Log.d(TAG, "Widget enabled")
    }
    
    override fun onDisabled(context: Context) {
        Log.d(TAG, "Widget disabled")
    }
}